klient/wait/conditions/conditions.go

```diff
@@ -58,6 +60,120 @@ func (c *Condition) ResourceScaled(obj k8s.Object, scaleFetcher func(object k8s.
+func (c *Condition) ResourceListMatchN(list k8s.ObjectList, n int, matchFetcher func(object k8s.Object) bool, listOptions ...resources.ListOption) apimachinerywait.ConditionFunc {
+	return func() (done bool, err error) {
+		if err := c.resources.List(context.TODO(), list, listOptions...); err != nil {
+			return false, nil
+		}
+		var found int
+		metaList, err := meta.ExtractList(list)
+		if err != nil {
+			return false, err
+		}
+		for _, obj := range metaList {
+			if o, ok := obj.(k8s.Object); ok && matchFetcher(o) {
+				found++
+			} else if !ok {
+				return false, fmt.Errorf("unexpected type %T in list, does not satisfy k8s.Object", obj)
+			}
+		}
+		return found >= n, nil
+	}
+}
+
+// ResourcesFound is a helper function that can be used to check for a set of objects. This function accepts a list
+// of named objects and will wait until it is able to retrieve each.
+func (c *Condition) ResourcesFound(list k8s.ObjectList) apimachinerywait.ConditionFunc {
+	return c.ResourcesMatch(list, func(object k8s.Object) bool { return true })
+}
+
+// ResourcesMatch is a helper function that can be used to check for a set of objects. This function accepts a list
+// of named objects and a match function, and will wait until it is able to retrieve each while passing the match validation.
+func (c *Condition) ResourcesMatch(list k8s.ObjectList, matchFetcher func(object k8s.Object) bool) apimachinerywait.ConditionFunc {
+	metaList, err := meta.ExtractList(list)
+	if err != nil {
+		return func() (done bool, err error) { return false, err }
+	}
+	objects := make(map[k8s.Object]bool)
+	for _, o := range metaList {
+		obj, ok := o.(k8s.Object)
+		if !ok {
+			panic(fmt.Errorf("unexpected type %T in list, does not satisfy k8s.Object", obj))
+		}
+		if obj.GetName() != "" {
+			objects[obj] = false
+		}
+	}
+	return func() (done bool, err error) {
+		found := 0
+		for obj, created := range objects {
+			if !created {
+				if err := c.resources.Get(context.TODO(), obj.GetName(), obj.GetNamespace(), obj); errors.IsNotFound(err) {
+					continue
+				} else if err != nil {
+					return false, err
+				}
+				if !matchFetcher(obj) {
+					continue
+				}
+			}
+			objects[obj] = true
+			found++
+		}
+		return len(objects) == found, nil
+	}
+}
+
+// ResourcesDeleted is a helper function that can be used to check for if a set of objects has been deleted. This function
+// accepts a list of named objects and will wait until it is not able to find each.
+func (c *Condition) ResourcesDeleted(list k8s.ObjectList) apimachinerywait.ConditionFunc {
+	metaList, err := meta.ExtractList(list)
+	if err != nil {
+		return func() (done bool, err error) { return false, err }
+	}
+	objects := make(map[k8s.Object]bool)
+	for _, o := range metaList {
+		obj, ok := o.(k8s.Object)
+		if !ok {
+			panic(fmt.Errorf("unexpected type %T in list, does not satisfy k8s.Object", obj))
```

Comment:
I don't think this should panic, just an error return should be ok.

Reply:
That would throw off the usage of condition slightly, if it returned (apimachinerywait.ConditionFunc, error) compared to other functions.

Or perhaps I could emit the error in the `ConditionFunc`?

(will address this in ResourcesMatch too)

Reply:
I am leaning towards returning a ConditionFunc that simply exits with the appropriate error

Reply:
@crandles Yes, that's what I meant, return ConditionFunc with an error, you are already doing it in other parts of the code.